src/main/scala/com/cognite/sdk/scala/v1/dataTypes.scala

```diff
@@ -10,4 +10,13 @@ sealed trait CogniteId
 final case class CogniteExternalId(externalId: String) extends CogniteId
 final case class CogniteInternalId(id: Long) extends CogniteId
 
-final case class TimeRange(min: Instant, max: Instant)
+// min and max needs to optional, since one of them can be provided alone.
+final case class TimeRange(min: Option[Instant] = None, max: Option[Instant] = None)
+final case class ConfidenceRange(min: Option[Double] = None, max: Option[Double] = None)
+
+// Used for filtering by label, labels: {containsAny: [{externalId: "label1"}, {externalId: "label2}]}
+// or labels: {containsAll: [{externalId: "label1"}, {externalId: "label2}]}
+final case class LabelContainsFilter(
+    containsAny: Option[Seq[CogniteExternalId]] = None,
+    containsAll: Option[Seq[CogniteExternalId]] = None
```

Comment:
This is great, and the right way to go! Do we know if the API accepts both `containsAny` and `containsAll` in a single request, or should we make a class that can be either `containsAny` OR `containsAll`?
Where is this documented btw? API docs just show `containsAny`.

Reply:
Yes, it is not shown in the API reference but It was a parameter I was using in my other project and available in the API. We will only have one of them, I tried creating two case classes for them and using Either but looks like Encoding for Either is a problem.

Reply:
Right! We should be able to do this with a trait, similar to what we do for the Ids
```
sealed trait CogniteId

final case class CogniteExternalId(externalId: String) extends CogniteId
final case class CogniteInternalId(id: Long) extends CogniteId
```

Reply:
Encoder became a problem with this again, I added encoders for case classes and defined the labels field as:  labels: Option[LabelContainsFilter] = None in the RelationshipsFilter and AssetsFilter. Is there something I am missing here?
. 
`
sealed trait LabelContainsFilter

final case class ContainsAny(containsAny: Seq[CogniteExternalId]) extends LabelContainsFilter
final case class ContainsAll(containsAll: Seq[CogniteExternalId]) extends LabelContainsFilter
`

Reply:
Not sure, which kind of problem are you seeing? "No implicit of type _"?

Reply:
So if I define an implicit encoder for LabelContainsFilter, ContainsAny and ContainsAll, this happens:
```
Could not identify primary constructor for com.cognite.sdk.scala.v1.LabelContainsFilter
[error]     deriveEncoder[LabelContainsFilter]
```
If I only define ContainsAny and ContainsAll implicit Encoders, this happens:

`Could not find io.circe.Encoder instance for Option[com.cognite.sdk.scala.v1.LabelContainsFilter]`

Reply:
Strange, seems to work for the Ids, see `common/package.scala`:
```
  implicit val cogniteIdItemsEncoder: Encoder[Items[CogniteId]] = deriveEncoder
  implicit val cogniteInternalIdEncoder: Encoder[CogniteInternalId] = deriveEncoder
  implicit val cogniteInternalIdItemsEncoder: Encoder[Items[CogniteInternalId]] = deriveEncoder
  implicit val cogniteExternalIdEncoder: Encoder[CogniteExternalId] = deriveEncoder
  implicit val cogniteExternalIdItemsEncoder: Encoder[Items[CogniteExternalId]] = deriveEncoder
```

Reply:
But I notice it doesn't derive an `Encoder[CogniteId]`, only the `Encoder[Items[CogniteId]]`.. perhaps we can do the same with `Encoder[Option[LabelContainsFilter]]` and `Encoder[ContainsAll/ContainsAny]`.

Reply:
I still could not manage to do this. The current version works but I want to do this proper way. It would be great id we take a look at this together when you have time.

Reply:
Turns out circe doesn't like encoding traits, and we have handled this before in the common/package.scala file. Have a look at the cogniteIdEncoder and see if we can do the same for our `LabelContainsFilter`, I mean this version:
```
sealed trait LabelContainsFilter

final case class ContainsAny(containsAny: Seq[CogniteExternalId]) extends LabelContainsFilter
final case class ContainsAll(containsAll: Seq[CogniteExternalId]) extends LabelContainsFilter
```

Reply:
Yes I had a look at that but it failed without 
```suggestion
  implicit val cogniteIdEncoder: Encoder[CogniteId] = new Encoder[CogniteId] {
    final def apply(i: CogniteId): Json = Json.obj(
      i match {
        case i: CogniteInternalId => ("id", Json.fromLong(i.id))
        case e: CogniteExternalId => ("externalId", Json.fromString(e.externalId))
      }
    )
  }
```
It did not work without this. I thought maybe we also need this but It was problematic to adapt this function, I could not manage to find correct Json function here.

Reply:
Should be something like
```
  implicit val labelContainsEncoder: Encoder[LabelContainsFilter] = new Encoder[LabelContainsFilter] {
    final def apply(i: LabelContainsFilter): Json = Json.obj(
      i match {
        case ca: ContainsAny => ("containsAny", Json.fromSeq??(ca.containsAny))
        case cAll: ContainsAll => ("containsAll", Json.fromSeq??(ca.containsAll))
      }
    )
  }
```

Reply:
Haha right, just ahead of me there. We have a (for example) `containsAny: Seq[String]` and we want to turn it into a
`Json.array( Json.object ( "externalId": "someString" ) )`.
Json arrays we can create with `Json.arr`, and objects like above with `Json.obj`.
I have meetings until 12.15, so here's the spoiler:
```
case conAll: ContainsAll => Json.arr(conAll.containsAll.map(s => Json.obj(("externalId", Json.fromString(s.externalId))) ):_* )
```
The tricky part is Json.arr takes a v`values: Json*`, i.e any number of Jsons.. in Scala we can pass this as `Seq[] :_*`, `:_*` is known as the splat operator: https://alvinalexander.com/bookmarks/scala/scalas-missing-splat-operator/ and takes us from a Sequence to "any number of parameters passed to function". This is the same concept as the * in python: https://stackoverflow.com/questions/2322355/proper-name-for-python-operator